pep-9999.rst

```diff
@@ -0,0 +1,1176 @@
+     separate ``sysconfig`` schemes.
+
+   This is a behavior change, but it seems to be defensible,
+   in that if your ``PYTHONHOME`` is a straight copy of the
+   distribution's Python, it should behave like the
+   distribution's Python.
+
+   If the ``PYTHONHOME`` is separate from the distribution,
+   such as a copy of the standard library from an unmodified
+   upstream Python (but used with a compatible interpreter
+   from the distribution), then the behavior is just like an
+   unmodified upstream Python (case 1) and does not change:
+
+   * Can you ``pip install`` without an override? Currently
+     yes; after this PEP yes, because there is no marker
+     file.
+
+   * Can it delete externally-installed packages (assuming
+     there is some well-defined concept of
+     "externally-installed packages" in this context)?
+     Currently yes; after this PEP yes, because the
+     distribution's changes to ``sysconfig`` won't be
+     visible, and there will only be one shared scheme.
+
+Specification
+=============
+
+Marking an interpreter as using an external package manager
+-----------------------------------------------------------
+
+Before a Python-specific package installer (that is, a tool such as
+pip - not an external tool such as apt) installs a package
+into a certain Python context, it should make the following
+checks by default:
+
+1. Is it running outside of a virtual environment? It can
+   determine this by whether ``sys.prefix ==
+   sys.base_prefix`` (but see `Backwards Compatibility`_).
+
+2. Is there a ``EXTERNALLY-MANAGED`` file in the directory
+   identified by ``sysconfig.get_path("stdlib",
+   sysconfig.get_default_scheme())``
+
+If both of these conditions are true, the installer should
+exit with an error message indicating that package
+installation into this Python interpreter's directory are
+disabled outside of a virtual environment.
+
+The installer should have a way for the user to override
+these rules, such as a command-line flag
+``--break-system-packages``. This option should not be
+enabled by default and should carry some connotation that
+its use is risky.
+
+The ``EXTERNALLY-MANAGED`` file is a metadata file in the
+`packaging core metadata format`_, which is an
+email-message-like format with headers and a body.  (At
+current writing, that format is defined exactly as what the
+standard library ``email.parser`` module can parse using
+``policy=email.policy.compat32``.) If the file can be parsed
+as a core metadata file, then the installer should output an
+error message from that file as part of its error. If
+``locale.getlocale(locale.LC_MESSAGES)`` returns
+non-``None`` and the first element is a string of the form
+``xx_YY``, and the file contains a header variable
+``Error-xx_YY`` or failing that ``Error-xx``, then the
+installer should use the value of that header as the error.
+Otherwise, it should use the body of the message as an
+error.
```

Comment:
We were going to do either TOML or "INI as parsed by ConfigParser" but @pradyunsg expressed a preference to use the email-ish format for now to avoid having to think about TOML standardization yet. Search Discord for "closer to METADATA" and you'll see the discussion.

I don't mind INI. Is it popular enough to use it? It feels like it's not used much in the packaging space, but I guess there's pydistutils.cfg and setup.cfg and pip.conf....

Anyway, I think once TOML is in the standard library we should switch. Perhaps that requires specifying that here. Maybe something like "Or, if a file `externally-managed.toml` exists, treat it like an `EXTERNALLY-MANAGED` file is present, and this PEP does not currently specify the error message behavior."

Reply:
venv uses a subset of the INI syntax (for ease of parsing with C code without dependencies). Although `EXTERNALLY-MANAGED` probably don’t need to be parsed in environments *that* constrained, I prefer INI-like to both email-like and TOML for similar reasons. This file does not need anything past the most basic construct: simple string—even booleans are probably not needed. So I perfer keeping things as simple as possible, maybe even simpler than INI: each key-value pair is on a line, separated by `=` and stripped of ' \n\r'.

Reply:
Honestly, any of the three formats we've talked about here will work for pip, because it already has the ability to read all three of the formats. And, the specific choice doesn't really affect distro maintainers much, since I expect they'd mostly add this in via a patch.

I suggested following METADATA / PKG-INFO / WHEEL file formats because anything that needs to read this file will definitely have the reader for that format already. Otherwise, _every_ wheel installer tool will need to know how to read from TOML (tricky, with the current state of the Python TOML ecosystem) or use a specific configuration of the standard library config parser (which is less annoying thanks to Python 2 being dead).

Reply:
Ah, for that custom `key = value` style file, that is modeled after pyvenv.cfg, that works too. pip has to read that file today, and we can just reuse that code. 🙃

Basically, anything we pick here works for pip. _I_ think following the email-like format makes it so that other future wheel installation tools (you know, like the ones I'm writing) don't need to know about additional file formats.

Reply:
Yeah, it occurred to me that `installer` is affected here (and I snuck in a link to it :) ), and also that `setup.py` is affected - `sudo ./setup.py install` also ought to refuse to run by default. So we'll need something that both installer and setuptools are going to be happy with.

I agree that any of the formats are fine - I want to nail down the semantics of what to do about the file if it exists, and then we can deal with the detail of the file format.

Reply:
From the other discussion about parsing and newlines, I changed this to INI-as-parsed-by-ConfigParser, but if it's useful to constrain the syntax for non-Python parsers I'm happy to do that.

(It seems convenient to make sure the file is still parsable by ConfigParser, I think, but I'll happily defer to implementors.)

Reply:
I think this is still open - do we want to constrain the file format any further for the sake of non-Python parsers, or is "INI-as-parsed-by-ConfigParser" sufficient?

Reply:
Let’s use INI-as-parsed-by-ConfigParser (I think that’s the conclusion in the call). If we add key-values in the file in the future, we should make the keys `all-lower-case` for compatibility (many INI parsers mess this up so we should be forgiving if we can).